### ISSUE TITLE
extending API

### ISSUE THREAD

**Issue by freimanas** [YOURLS/YOURLS#1873]
Hi,

I was looking into making a plugin to extend functionality of the API.
I would like to be able to pass additional variable "prefix" on creating short url, instead making of the same prefix for each link the YOURLS creates I would like it to be specific for that API call.

example:
`/yourls-api.php?signature=xxxxxxx&action=shorturl&url=http://mydomain.com/content&prefix=abc_&format=json`

would create a short link: abc_[original_short_url]

I just could not work out if I can add that, as cant see any possibility to check for that additional parameter in `function yourls_api_action_shorturl()`

am I missing something here? any help would be much appreciated.

thanks

**Comment by freimanas** [comment]
by the way I did RTFM and none of the hooks seems to do what I need here.. 

or at least "I think" none does.

**Comment by ozh** [comment]
You could do something like this:
```php
// highjack regular "shorturl" action
yourls_add_filter( 'api_actions', 'freimanas_custom_action' );

function freimanas_custom_action( $api_actions ) {
    $api_actions['shorturl'] = 'freimanas_api_action_shorturl';
    return $api_actions;
}

// define new "shorturl" action
function freimanas_api_action_shorturl( ) {
    // steal what being done in includes/functions-api.php#L17
    // and modify it to take $_REQUEST['prefix'] into account
}
```

Another way could be to limit yourself to filters provided by function `yourls_add_new_link()` in the main `functions.php` and simply check if it's an API request  :
```php
yourls_add_filter( 'random_keyword', 'freimanas_prefix_keyword' );
yourls_add_filter( 'custom_keyword', 'freimanas_prefix_keyword' );

function freimanas_prefix_keyword( $keyword )  {
    if( defined( 'YOURLS_IS_API' ) && YOURLS_IS_API ) {
        $prefix = ( isset( $_REQUEST['prefix '] ) ? $_REQUEST['prefix '] : '' );
        $keyword = $prefix . $keyword;
    }

    return $keyword;
}
```

HTH.

**Comment by freimanas** [comment]
I have used the second approach (well tried 1st for a few hours but with no luck).

second approach is quite nice, BUT!
`if( defined( 'YOURLS_IS_API' ) && YOURLS_IS_API )`
always failed check, not sure why, but by removing this would still work as it does not matter if API as $prefix would be '' and would not change anything.

so by removing check on YOURLS_IS_API I made it working.
Also needed to change functions.php to sanitize after I change the $keyword as if its made via API its possible to have malicious characters so I ended up receiving one link from API and it was not working (with underscores) and by removing underscores - all worked fine.

well thanks for a great tool!
